tests/lib/hardware/libvirt.py

```diff
@@ -0,0 +1,300 @@
+                        f"ssh on {ip}")
+
+    def _get_ips(self, timeout=45):
+        """get the ip addresses of the guest domain from the DHCP leases"""
+        ips_found = []
+        xmldoc = minidom.parseString(self._dom.XMLDesc())
+        mac_list = xmldoc.getElementsByTagName('mac')
+        stop = datetime.datetime.now() + datetime.timedelta(seconds=timeout)
+        logger.info(f"node {self.name}: wait {timeout}s to get IP address")
+        while len(mac_list) and datetime.datetime.now() < stop:
+            for mac in mac_list:
+                mac_addr = mac.attributes["address"].value
+                for d in self._network.DHCPLeases():
+                    if d['mac'].lower() == mac_addr.lower():
+                        ips_found.append(d['ipaddr'])
+            if len(ips_found):
+                logger.info(f"node {self.name}: found IPs {ips_found}")
+                return ips_found
+            time.sleep(3)
+        raise Exception(f"node {self.name}: no IP address found")
+
+    def _backing_file_create(self):
+        # TODO(toabctl): move the temp image to a different tmp dir
+        if os.path.exists(self._snap_img_path):
+            logger.info(f"node {self.name}: Delete available backing image "
+                        f"{self._snap_img_path}")
+            os.remove(self._snap_img_path)
+        subprocess.check_call(f"qemu-img create -f qcow2 -F qcow2 -o "
+                              f"backing_file={config.PROVIDER_LIBVIRT_IMAGE} "
+                              f"{self._snap_img_path} 10G",
+                              shell=True)
+        logger.info(f"node {self.name}: created qcow2 backing file under"
+                    f"{self._snap_img_path}")
+
+    def _cloud_init_seed_create(self):
+        user_data = """#cloud-config
+debug: True
+ssh_authorized_keys:
+- {}
+        """
+        meta_data = """---
+instance-id: {}
+local-hostname: {}
+        """
+
+        iso_cmd = shutil.which('mkisofs')
+        if not iso_cmd:
+            raise Exception('mkisofs command not found')
+
+        if os.path.exists(self._cloud_init_seed_path):
+            os.remove(self._cloud_init_seed_path)
+        with tempfile.TemporaryDirectory() as tempdir:
+            with open(os.path.join(tempdir, 'user-data'), 'w') as ud:
+                ud.write(user_data.format(self._ssh_public_key))
+            with open(os.path.join(tempdir, 'meta-data'), 'w') as md:
+                md.write(meta_data.format(uuid.uuid4(), self.name))
+            # create the seed file
+            args = [iso_cmd,
+                    '-output', self._cloud_init_seed_path,
+                    '-volid', 'cidata',
+                    '-joliet', '-rock',
+                    tempdir]
+            subprocess.check_call(args, stdout=subprocess.DEVNULL,
+                                  stderr=subprocess.DEVNULL)
+
+    def _get_domain(self, domain_name, image, cloud_init_seed, network_name,
+                    memory):
+        return """
```

Comment:
I'm not a big fan of inline file within code.
Maye you could split the xml to an external file ?

Reply:
yeah. same here. I wonder if I should just use a jinja template for that. But that needs some more work I think.
Are you fine with doing that in a follow up PR?

Reply:
We can also use dedent to make it much more readable.

Reply:
Fixed in #55